java/clients/src/main/java/sleeper/clients/teardown/TerminateEMRClusters.java

```diff
@@ -0,0 +1,121 @@
+ * limitations under the License.
+ */
+
+package sleeper.clients.teardown;
+
+import com.amazonaws.services.elasticmapreduce.AmazonElasticMapReduce;
+import com.amazonaws.services.elasticmapreduce.AmazonElasticMapReduceClientBuilder;
+import com.amazonaws.services.elasticmapreduce.model.ClusterState;
+import com.amazonaws.services.elasticmapreduce.model.ClusterSummary;
+import com.amazonaws.services.elasticmapreduce.model.ListClustersRequest;
+import com.amazonaws.services.elasticmapreduce.model.TerminateJobFlowsRequest;
+import com.amazonaws.services.s3.AmazonS3;
+import com.amazonaws.services.s3.AmazonS3ClientBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import sleeper.configuration.properties.InstanceProperties;
+import sleeper.util.PollWithRetries;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static java.lang.Math.min;
+import static sleeper.configuration.properties.UserDefinedInstanceProperty.ID;
+import static sleeper.configuration.properties.UserDefinedInstanceProperty.OPTIONAL_STACKS;
+
+public class TerminateEMRClusters {
+    private static final Logger LOGGER = LoggerFactory.getLogger(TerminateEMRClusters.class);
+    private static final long POLL_INTERVAL_MILLIS = 30000;
+    private static final int MAX_POLLS = 30;
+
+    private final PollWithRetries poll = PollWithRetries.intervalAndMaxPolls(POLL_INTERVAL_MILLIS, MAX_POLLS);
+
+    private final AmazonElasticMapReduce emrClient;
+    private final String clusterPrefix;
+    private static final List<String> ACTIVE_STATES = List.of(ClusterState.STARTING.name(), ClusterState.BOOTSTRAPPING.name(),
+            ClusterState.RUNNING.name(), ClusterState.WAITING.name(), ClusterState.TERMINATING.name());
+
+    public TerminateEMRClusters(AmazonElasticMapReduce emrClient, InstanceProperties properties) {
+        this.emrClient = emrClient;
+        this.clusterPrefix = "sleeper-" + properties.get(ID) + "-";
+    }
+
+    public void run() throws InterruptedException {
+        List<ClusterSummary> clusters = emrClient.listClusters(new ListClustersRequest()
+                .withClusterStates(ACTIVE_STATES)).getClusters();
+        List<String> clusterIds = clusters.stream()
+                .filter(cluster -> cluster.getName().startsWith(clusterPrefix))
+                .map(ClusterSummary::getId)
+                .collect(Collectors.toList());
+        if (clusterIds.size() == 0) {
+            LOGGER.info("No running clusters to terminate");
+        } else {
+            LOGGER.info("Terminating {} running clusters", clusterIds.size());
+            terminateClusters(clusterIds);
+            LOGGER.info("Waiting for clusters to terminate");
+            pollUntilTerminated();
+        }
+    }
+
+    private void terminateClusters(List<String> clusters) {
+        // Can only terminate 10 clusters at a time
+        // See https://docs.aws.amazon.com/emr/latest/APIReference/API_TerminateJobFlows.html
```

Comment:
Also see https://docs.aws.amazon.com/general/latest/gr/emr.html - there is a limit on 10 calls to this API per second. If each call terminates 10 then that's 100 per second, which we probably won't hit. But given that terminating clusters is a slow operation anyway (order of minutes) then maybe we should sleep for 1 second after every call to emrClient.terminateJobFlows for safety?

Reply:
It looks like the sustained rate (based on the replenish rate) is 1 request every 2 seconds. I've updated the code to wait for 5 seconds between requests to be on the safe side.